### ISSUE TITLE
[Attention] Support for post-softmax fp8 scaling

### ISSUE THREAD

**Issue by dan-garvey** [iree-org/iree#20823]
### Request description

here's a brief code block that shows what we want to accomplish, using the simplified "decomposed" version of attention (ignore type casting):
```
            # Q @ V.T
            attn_weights = ops.matmul(
                q.to(torch.float32), k.transpose(2, 3).to(torch.float32)
            )
            attn_weights = attn_weights / math.sqrt(self.attn_head_dim)

            # Apply attention mask.
            attn_weights = attn_weights + mask

            # Softmax
            attn_weights = ops.softmax(
                ops.to(attn_weights, dtype=torch.float32), dim=-1
            )

            # This is the changed part!
            # Scale the softmax output to a range of values representable in fp8
            attn_weights = (
                        probs_quantizer.quantize(attn_weights)
            )

            attn_weights = ops.to(attn_weights, dtype=q.dtype)
            outputs = ops.matmul(attn_weights, v) 

            # this is also a change
            outputs = outputs/probs_quantizer.scale
            
             return outputs

```

I went ahead and wrote comments above the parts that are different from the existing kernel. 

This quantization is already implemented successfully in rocvllm. So you don't have to read triton spaghetti, here is a summary of what they are doing:

```
         # scale sm_scale by log_2(e) and use 2^x in the loop
         qk_scale = sm_scale * 1.44269504089
         qk_scale *= q_scale * k_scale
         acc_scale = p_scale * v_scale

         qk = q @ k
         qk *= qk_scale 
         nmax = max(max, qk)
         qk = qk - nmax
        
         p = exp2(qk)
         p *= p_descale -> cast to fp8
         acc = p @ v
         acc *= acc_scale (p_scale)
         acc *= o_descale (I assume this is the input quant scale from the next op)
         acc = clip(acc)

```

To try to accomplish this, a few people helped build this branch:
https://github.com/iree-org/iree/compare/main...users/dan-garvey/warplogic_edits

however there are a few problems with this branch as is. First is that the prob quantizer is being applied before the softmax, which will not preserve its effect, because softmax output is always between 0 and 1.  However, there is a comment in the code that suggests that its fine where it is. (https://github.com/iree-org/iree/compare/main...KyleHerndon:iree:warplogic#diff-53b5f9e46161107dea58707330f1a761054e5021bb39506d49d37ad05260d2f8L374)
The other problem is that the descaling step is not applied. (We can do this in python though, as it can just be applied to the output)

Because that branch had a lot of changes, and the prob_output_scale is actually fixed for all layers, I was planning to work off of a branch that just hardcodes it to avoid that. 

I can provide whatever ir or golden i/o is required, but I think this might just be a math issue.

**Comment by pashu123** [comment]
@dan-garvey I can't see the bold parts that you wanted to highlight. Maybe something is missing.

**Comment by dan-garvey** [comment]
can't bold in code blocks apparently

            **# Scale the softmax output to a range of values representable in fp8
            attn_weights = (
                        probs_quantizer.quantize(attn_weights)
            )**


            **outputs = outputs/probs_quantizer.scale**



 p *= p_descale -> cast to fp8

acc *= acc_scale (p_scale)

**Comment by MaheshRavishankar** [comment]
cc @IanWood1  and @Groverkss  if they have comments.

**Comment by dan-garvey** [comment]
I got some math help. I think that the right place to do the scaling is in or after this function:
https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp#L574


and that the comment here is actually just incorrect or refers to when you want to do softmax itself in fp8, as otherwise it mathematically has no affect on the output of softmax. https://github.com/iree-org/iree/blob/main/compiler/src/iree/compiler/Dialect/LinalgExt/IR/AggregatedOpInterfaceImpl.cpp#L374

What I don't know is where all the "get" functions are coming from and whether or not I can just add an elementwiseInPlaceOp to multiply P by the scale.

**Comment by pashu123** [comment]
The first code looks more like decomposed attention (i.e., without any tiling), while the other one is a tiled implementation (i.e., flash attention). I don't know whether we are just allowed to multiply P by scale in the tiled implementation. I haven't touched the AggregatedOpInterface of the operation, but it looks like a tiled Implementation.  @Groverkss do you have any comments?

**Comment by MaheshRavishankar** [comment]
> The first code looks more like decomposed attention (i.e., without any tiling), while the other one is a tiled implementation (i.e., flash attention). I don't know whether we are just allowed to multiply P by scale in the tiled implementation. I haven't touched the AggregatedOpInterface of the operation, but it looks like a tiled Implementation. [@Groverkss](https://github.com/Groverkss) do you have any comments?

I dont think there is anything w.r.t to Tiled implementation in the `AggregateOpInterface` is as its name "should" imply. You are taking an operation and just replacing it with its components...
If you need any information while decomposing the aggregate op you need to add it to the operation. 

I must say I still dont get a handle on what needs to happen here. I thought this was a matter of adding an operand to the operation.

**Comment by pashu123** [comment]
Yes, AggregateOpInterface has nothing to do with Tiling, but it's generating what (innermost loop of) flash attention would generate as opposed to simple decomposition. I am confused about mapping them.

**Comment by dan-garvey** [comment]
> Yes, AggregateOpInterface has nothing to do with Tiling, but it's generating what (innermost loop of) flash attention would generate as opposed to simple decomposition. I am confused about mapping them.

I think it must given all the "get" methods used to pull data that would otherwise be local to the op

**Comment by dan-garvey** [comment]
@Groverkss confirmed that scaling P in the location I mentioned should be fine even given that it is tiled. 

I'm still a bit out of my neck of the woods here, can someone explain why the affine maps for the following are wrong?
```
  Value probScale = b.create<arith::ConstantOp>(loc, b.getFloatAttr(b.getF32Type(), 224.8484));
  p = elementwiseValueInPlace<arith::MulFOp>(b, loc, pMap, sMap, p, probScale);

error: 'linalg.generic' op expected operand rank (0) to match the result rank of indexing_map #0 (3)
```

edit:: sMap == pMap is earlier and the code and makes no sense to me

edit:: edit:: sMap is for S not scale, had to fetch the code to generate a map for scale from elsewhere, problem solved

**Comment by MaheshRavishankar** [comment]
@dan-garvey it took me a while, but I think I understand what you are trying to do. IIUC the current attention op cannot handle this. Still a couple of questions

```
attn_weights = (
                        probs_quantizer.quantize(attn_weights)
            )
```

what is exactly the computation in `probs_quantizer.quantize(attn_weights)`.

In all probability we need to modify the attention op for this to work.

**Comment by dan-garvey** [comment]
There are two parts, one is supporting this argument which is addressed largely by the warplogic branch, and the other is getting the math right. I'd like to get the math right first, and then land support for the argument

probs_quantizer.quantize is just taking the softmax output and changing it from range (0,1) to something more appropriate for the fp8 datatype. 

This branch shows my current progress on this. https://github.com/iree-org/iree/compare/users/dan-garvey/fp8_attn_kernel_dev?expand=1

This does not result in correct numerics, but it is better than what we had before (all 0 output). This branch gives output with some nans in fp8, and just bad numerics in fp32. My current hypothesis is that scaling P after it is used to calculate newSum is throwing off the calculation of the fused P @ V + newAcc generic. 


CC: @IanWood1

**Comment by MaheshRavishankar** [comment]
> There are two parts, one is supporting this argument which is addressed largely by the warplogic branch, and the other is getting the math right. I'd like to get the math right first, and then land support for the argument
> 
> probs_quantizer.quantize is just taking the softmax output and changing it from range (0,1) to something more appropriate for the fp8 datatype.
> 
> This branch shows my current progress on this. https://github.com/iree-org/iree/compare/users/dan-garvey/fp8_attn_kernel_dev?expand=1
> 
> This does not result in correct numerics, but it is better than what we had before (all 0 output). This branch gives output with some nans in fp8, and just bad numerics in fp32. My current hypothesis is that scaling P after it is used to calculate newSum is throwing off the calculation of the fused P @ V + newAcc generic.
> 
> CC: [@IanWood1](https://github.com/IanWood1)

@dan-garvey  did you verify numerics with the decomposed attention. I think you did, but just confirming.

**Comment by dan-garvey** [comment]
> > There are two parts, one is supporting this argument which is addressed largely by the warplogic branch, and the other is getting the math right. I'd like to get the math right first, and then land support for the argument
> > probs_quantizer.quantize is just taking the softmax output and changing it from range (0,1) to something more appropriate for the fp8 datatype.
> > This branch shows my current progress on this. https://github.com/iree-org/iree/compare/users/dan-garvey/fp8_attn_kernel_dev?expand=1
> > This does not result in correct numerics, but it is better than what we had before (all 0 output). This branch gives output with some nans in fp8, and just bad numerics in fp32. My current hypothesis is that scaling P after it is used to calculate newSum is throwing off the calculation of the fused P @ V + newAcc generic.
> > CC: [@IanWood1](https://github.com/IanWood1)
> 
> [@dan-garvey](https://github.com/dan-garvey) did you verify numerics with the decomposed attention. I think you did, but just confirming.

In pytorch with fake quant yes. I'm not sure if decomposed compiled via iree works.
